xds/src/main/java/io/grpc/xds/LoadReportClient.java

```diff
@@ -97,35 +92,39 @@
    * no-op.
    */
   void startLoadReporting() {
-    if (started) {
-      return;
+    synchronized (lock) {
+      if (started) {
+        return;
+      }
+      started = true;
+      logger.log(XdsLogLevel.INFO, "Starting load reporting RPC");
+      startLrsRpc();
     }
-    started = true;
-    logger.log(XdsLogLevel.INFO, "Starting load reporting RPC");
-    startLrsRpc();
   }
 
   /**
    * Terminates load reporting. Calling this method on an already stopped
    * {@link LoadReportClient} is no-op.
    */
   void stopLoadReporting() {
-    if (!started) {
-      return;
-    }
-    logger.log(XdsLogLevel.INFO, "Stopping load reporting RPC");
-    if (lrsRpcRetryTimer != null) {
-      lrsRpcRetryTimer.cancel();
-    }
-    if (lrsStream != null) {
-      lrsStream.close(Status.CANCELLED.withDescription("stop load reporting").asException());
+    synchronized (lock) {
+      if (!started) {
+        return;
+      }
+      started = false;
+      logger.log(XdsLogLevel.INFO, "Stopping load reporting RPC");
+      if (lrsRpcRetryTimer != null) {
+        lrsRpcRetryTimer.cancel(false);
+      }
```

Comment:
There is a race condition: 
- `stopLoadReporting()` is called and holds the lock.
- `LoadReportingTask.run()` is called, trying to acquire the lock, and being blocked because the lock can not be acquired.
- `stopLoadReporting()` turns `started = false`, and then calls `lrsRpcRetryTimer.cancel(false)` which does nothing because `LoadReportingTask.run()` is already running and still trying to acquire the lock.
- `stopLoadReporting()` returns and releases the lock.
- `LoadReportingTask.run()` finally acquires the lock and calls `startLrsRpc()`.

Reply:
I've considered this. When `stopLoadReporting()` is releasing the lock, `started` should have been set to `false`. The first statement of `startLrsRpc()` checks that and return immediately if already stopped.

Reply:
Oh, the method name `startLrsRpc()` confused me, it actually should be `restartLrsRpc()`.

Reply:
`start*` is fine, it is used both for the first time start and restart cases.